engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -1927,6 +1942,48 @@ static int CreateBuffer(lua_State* L)
 
     dmBuffer::HBuffer buffer = dmGameSystem::UnpackLuaBuffer(lua_buffer);
 
+    // Regardless of transferring ownership, if the lua buffer is referring to a resource,
+    // we nee to copy that buffer into this new resource
+    if (lua_buffer->m_Owner == dmScript::OWNER_RES)
+    {
+        uint32_t stream_count;
+        dmBuffer::GetNumStreams(buffer, &stream_count);
+        dmBuffer::StreamDeclaration* streams_decl = (dmBuffer::StreamDeclaration*) malloc(stream_count * sizeof(dmBuffer::StreamDeclaration));
+
+        for (uint32_t i = 0; i < stream_count; ++i)
+        {
+            dmhash_t stream_name;
+            dmBuffer::ValueType stream_type;
+            uint32_t stream_type_count;
+
+            dmBuffer::GetStreamName(buffer, i, &stream_name);
+            dmBuffer::GetStreamType(buffer, stream_name, &stream_type, &stream_type_count);
+            streams_decl[i].m_Name  = stream_name;
+            streams_decl[i].m_Type  = stream_type;
+            streams_decl[i].m_Count = stream_type_count;
+        }
+
+        uint32_t dst_count = 0;
+        dmBuffer::Result br = dmBuffer::GetCount(buffer, &dst_count);
+
+        dmBuffer::HBuffer dst_buffer;
+        br = dmBuffer::Create(dst_count, streams_decl, stream_count, &dst_buffer);
+        free(streams_decl);
+
+        if (br != dmBuffer::RESULT_OK) {
+            return luaL_error(L, "Unable to create copy buffer: %s (%d).", dmBuffer::GetResultString(br), br);
+        }
+
+        // Copy supplied data to buffer
+        br = dmBuffer::Copy(dst_buffer, buffer);
```

Comment:
I honestly feel that we should have another Copy function that does all this work.
Would you mind adding such a function?

Reply:
Yes I was thinking about that as well, I'll add it 👍